packages/merge-styles/src/IStyleSet.ts

```diff
@@ -1,5 +1,37 @@
 import { IStyle } from './IStyle';
+import { IStyleFunctionOrObject, IStyleFunction } from './IStyleFunction';
 
-export type IStyleSet = {
-  [key: string]: IStyle;
+export type Diff<T extends keyof any, U extends keyof any> = ({ [P in T]: P } &
+  { [P in U]: never } & { [x: string]: never })[T];
+
+export type Omit<U, K extends keyof U> = Pick<U, Diff<keyof U, K>>;
+
+/**
+ * A style set is a dictionary of display areas to IStyle objects.
+ * It may optionally contain style functions for sub components in the special `subComponentStyles`
+ * property.
+ */
+export type IStyleSet<TStyleSet extends IStyleSet<TStyleSet>> = {
+  [P in keyof Omit<TStyleSet, 'subComponentStyles'>]: IStyle
+} & {
+  subComponentStyles?: { [P in keyof TStyleSet['subComponentStyles']]: IStyleFunctionOrObject<any, IStyleSet<any>> };
+};
+
+/**
+ * A concatenated style set differs from `IStyleSet` in that subComponentStyles will always be a style function.
+ */
+export type IConcatenatedStyleSet<TStyleSet extends IStyleSet<TStyleSet>> = {
```

Comment:
There is no difference between this and IProcessedStyleSet... I'm thinking about whether we gain much in differentiating them. The difference is that mergestyleset has a side effect of registering them but concat doesn't. The result is the same type structurally.

Reply:
There is a difference, look closely. 

One contains classnames, one contains IStyles.

Reply:
This is not just specified in the code, but called out in the comments as well. :)